src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/MsBuildModelWatcher.cs

```diff
@@ -0,0 +1,63 @@
+﻿using Microsoft.Build.Evaluation;
+using Microsoft.VisualStudio.ProjectSystem.Utilities;
+using Microsoft.VisualStudio.ProjectSystem.VS.Utilities;
+using System.Threading.Tasks;
+using System.Threading;
+using System.ComponentModel.Composition;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Editor
+{
+    [Export(typeof(IMsBuildModelWatcher))]
+    internal class MsBuildModelWatcher : OnceInitializedOnceDisposedAsync, IMsBuildModelWatcher
+    {
+        private readonly IProjectThreadingService _threadingService;
+        private readonly IFileSystem _fileSystem;
+        private readonly IMsBuildAccessor _accessor;
+        private readonly UnconfiguredProject _unconfiguredProject;
+        private string _tempFile;
+
+        [ImportingConstructor]
+        public MsBuildModelWatcher(IProjectThreadingService threadingService, 
+            IFileSystem fileSystem, 
+            IMsBuildAccessor accessor, 
+            UnconfiguredProject unconfiguredProject) :
+            base(threadingService.JoinableTaskContext)
+        {
+            _threadingService = threadingService;
+            _fileSystem = fileSystem;
+            _accessor = accessor;
+            _unconfiguredProject = unconfiguredProject;
+        }
+
+        public void Initialize(string tempFile)
+        {
+            _tempFile = tempFile;
+            _threadingService.ExecuteSynchronously(() => InitializeAsync());
+        }
+
+        public void ProjectXmlHandler(object sender, ProjectXmlChangedEventArgs args)
```

Comment:
This is the part I'm concerned about from a synchronization standpoint, as I have to do this write from a background thread because I'm called holding a CPS lock, and I can't switch to the UI thread. @lifengl, if you could take a look at this it would be appreciated.